### PR TITLE
Remove visiblity and lookup deprecation (Take 2)

### PR DESCRIPTION
**Pull request by JinShil**
Reboot of https://github.com/dlang/dmd/pull/7241
Credit to @MartinNowak 

~** This is not ready yet **~

### CODE REVIEW

**Comment by thewilsonator** [comment]
cc @atilaneves 
https://github.com/kaleidicassociates/excel-d/blob/master/source/xlld/wrap/wrap.d#L604-L607

**Comment by atilaneves** [comment]
@thewilsonator Not entirely sure why I was cc'ed.

**Comment by wilzbach** [comment]
Because excel-d is failing with this PR:

```
source/xlld/wrap/wrap.d(607,24): Error: `test.d_funcs.shouldNotBeAProblem` is not visible from module `wrap`
```

**Comment by jacob-carlborg** [comment]
> ** This is not ready yet **

Next time try creating a draft PR: https://github.blog/2019-02-14-introducing-draft-pull-requests.

**Comment by thewilsonator** [comment]
> Because excel-d is failing with this PR:

Yep. Sorry I should have put that in the ping.

**Comment by JinShil** [comment]
I believe all the failures in buildkite are legitimate.  The only thing I have left (I think) is to improve some of the error messages and prevent duplicate error messages.

File: test/fail_compilation/test143.d
Comment (Geod24) [inline test/fail_compilation/test143.d]: That's a confusing error message

File: test/fail_compilation/fail313.d
Comment (Geod24) [inline test/fail_compilation/fail313.d]: "not accessible" should be eliminated methinks

File: test/fail_compilation/fail313.d
Comment (Geod24) [inline test/fail_compilation/fail313.d]: Also confusing: the compiler should either not suggest anything or suggest static import

File: test/fail_compilation/fail262.d
Comment (Geod24) [inline test/fail_compilation/fail262.d]: Awesome improvement, but looks unrelated to this ? Or does the error improve as a result of changing a code path?

File: test/fail_compilation/fail18979.d
Comment (Geod24) [inline test/fail_compilation/fail18979.d]: Shouldn't it be an "undefined identifier" error ?
Also, in a perfect world, we shouldn't show "__ctor" to the user (or any symbol which is not defined in code), but that was already the case before.

**Review by Geod24 — COMMENTED** [review]
Thanks for doing this. A few comments on error messages, but looks like you're still working on it.

File: test/fail_compilation/fail262.d
Comment (JinShil) [reply test/fail_compilation/fail262.d]: Yes, the error message changes due to a change in the code path.

File: test/fail_compilation/fail18979.d
Comment (JinShil) [reply test/fail_compilation/fail18979.d]: How about now?

File: test/compilable/checkimports1.d
Comment (Geod24) [inline test/compilable/checkimports1.d]: That's a duplicate of `diag12598` no ?

Comment (JinShil) [reply test/compilable/checkimports1.d]: Indeed it is.  Removed.  Thanks!

File: src/dmd/access.d
Comment (Geod24) [inline src/dmd/access.d]: Shouldn't this whole function be reworked as a result ?
Not necessarily now tho.

File: src/dmd/globals.d
Comment (Geod24) [inline src/dmd/globals.d]: 2010-05 is a bit past us

File: src/dmd/mars.d
Comment (Geod24) [inline src/dmd/mars.d]: I wish in the future we can issue those deprecations way before we actually make the switch useless. But this has been draging its feet for so long that I don't think it matters in this case.

File: src/dmd/traits.d
Comment (Geod24) [inline src/dmd/traits.d]: Wasn't that a bug ? Also do we still need `noaccesscheck`?

File: src/dmd/typesem.d
Comment (Geod24) [inline src/dmd/typesem.d]: This could just be:
```d
s = mt.sym.search(e.loc, ident, (sc.flags & SCOPE.ignoresymbolvisibility) | IgnorePrivateImports);
```

File: src/dmd/typesem.d
Comment (Geod24) [inline src/dmd/typesem.d]: Ditto about the `flags` variable being used only once

File: test/fail_compilation/checkimports2.d
Comment (Geod24) [inline test/fail_compilation/checkimports2.d]: Those comments don't make sense anymore. I would just put the FQN of the symbol being accessed if you want to keep a comment, so `.X` for the first and `imports.imp1.Y` for the second.

File: test/fail_compilation/diag5385.d
Comment (Geod24) [inline test/fail_compilation/diag5385.d]: Okay I didn't have my morning coffee yet, but those suggestions don't make much sense to me. Either we should say it's not visible, or just not suggest the invisible symbol in the suggestion.

File: test/fail_compilation/dip22a.d
Comment (Geod24) [inline test/fail_compilation/dip22a.d]: Same thing here about suggestion

File: src/dmd/access.d
Comment (JinShil) [reply src/dmd/access.d]: Yes, I'm hoping I can delete the entire function when I'm done.

File: src/dmd/globals.d
Comment (JinShil) [reply src/dmd/globals.d]: That's intentional.  It should be removed 1 year from when this is released so it no longer displays a deprecation warning.

File: src/dmd/traits.d
Comment (JinShil) [reply src/dmd/traits.d]: Wasn't what a bug?  I'll look into removing the `noaccesscheck`.

File: test/fail_compilation/diag5385.d
Comment (JinShil) [reply test/fail_compilation/diag5385.d]: I'm planning to remove the suggestion.

File: test/fail_compilation/checkimports2.d
Comment (JinShil) [reply test/fail_compilation/checkimports2.d]: I plan to remove suggestions for invisible symbols.

File: test/fail_compilation/dip22a.d
Comment (JinShil) [reply test/fail_compilation/dip22a.d]: I'm planning to remove suggestions for invisible symbols.

File: src/dmd/typesem.d
Comment (JinShil) [reply src/dmd/typesem.d]: I don't think so because `SCOPE.ignoresymbolvisibility` is `0x200` and `IgnorePrivateImports` is `0x80`.  If they were the same value, I think it would work, but as long as they're different, I think this is the way it needs to be done.  Please correct me if I'm wrong.

Comment (Geod24) [reply src/dmd/typesem.d]: Oh, didn't realize they were different enum 🤦‍♂️

File: src/dmd/access.d
Comment (JinShil) [reply src/dmd/access.d]: I couldn't get rid of it.  It is called for some safety checks, specifically https://github.com/dlang/dmd/blob/master/test/fail_compilation/test18554.d

File: test/fail_compilation/dip22a.d
Comment (JinShil) [reply test/fail_compilation/dip22a.d]: Removed

File: test/fail_compilation/diag5385.d
Comment (JinShil) [reply test/fail_compilation/diag5385.d]: Removed.

File: test/fail_compilation/checkimports2.d
Comment (JinShil) [reply test/fail_compilation/checkimports2.d]: Removed.

File: test/fail_compilation/fail313.d
Comment (JinShil) [reply test/fail_compilation/fail313.d]: Suggestion removed.

File: test/fail_compilation/test143.d
Comment (JinShil) [reply test/fail_compilation/test143.d]: Suggestion removed.

File: src/dmd/traits.d
Comment (JinShil) [reply src/dmd/traits.d]: I couldn't remove `noaccesscheck`.  If I do `runnable/traits.d` emits failures.

File: src/dmd/cli.d
Comment (wilzbach) [inline src/dmd/cli.d]:
```suggestion
        Feature.Deprecated deprecated_; /// whether the feature is still in use
```

**Comment by wilzbach** [comment]
dlang-bot update PR: https://github.com/dlang/dlang-bot/pull/229

**Comment by wilzbach** [comment]
The Azure failure is pretty weird:

```
"D:\a\1\s\generated\Windows\Release\Win32\dmd.exe" -conf= -g -m32mscoff -O -w -dip25 -I../druntime\import -unittest -c  -ofunittest2.obj std\range\package.d  std\range\primitives.d  std\range\interfaces.d
Error: out of memory
```

I guess we split up building Phobos even more for win32 :/

**Comment by rainers** [comment]
> The Azure failure is pretty weird:
> I guess we split up building Phobos even more for win32 :/

BTW: the autotester uses posix.mak to build the phobos unittests in parallel and per file compilation on windows, too. See target `auto-tester-test`.

**Comment by atilaneves** [comment]
I've fixed excel-d in the meanwhile.

**Comment by wilzbach** [comment]
> > The Azure failure is pretty weird:
> > I guess we split up building Phobos even more for win32 :/
> 
> BTW: the autotester uses posix.mak to build the phobos unittests in parallel and per file compilation on windows, too. See target `auto-tester-test`.

Yeah, but I guess the auto-tester machines have more memory?
Do you have a better idea/solution to fixing this out of memory error on Azure?

**Comment by rainers** [comment]
> Yeah, but I guess the auto-tester machines have more memory?

Might be, but single file compilation will likely need less memory and is what posix.mak does. If parallel builds are a problem due to memory resources, it could be used with JOBS=1.

**Comment by rainers** [comment]
> Do you have a better idea/solution to fixing this out of memory error on Azure?

This might also help: https://github.com/dlang/dmd/pull/9650

**Comment by JinShil** [comment]
I have done as much as I can with this.  There is one bug remaining:

Given the following files:
├── fail313.d
└── imports
    ├── a313.d
    └── b313.d

```D
module fail313;
import imports.a313;
void main()
{
    imports.b313.bug();
}
```
```D
module imports.a313;
private import imports.b313;
```
```D
module imports.b313;
void bug() { }
```
```
dmd fail313.d
fail313.d(b): Error: module `imports.b313` is not accessible here, perhaps add `static import imports.b313;`
```
There error is emitted from the call to `checkAccess` at the bottom of this code block:
https://github.com/dlang/dmd/blob/a147141de604db226c3240c78521786500fe5978/src/dmd/expressionsem.d#L10993-L11008

The problem is it should have never gotten that far.  The call to `ie.sds.search()` or the call to `symbolIsVisible` should have not returned `imports.b313`

It's an existing bug in symbol lookup/visibility that is being masked by the now deprecated access checks. I need to fix that, and remove the call to `checkAccess`.  I've been working on it for 2 days and can't figure it out.  I could use some help.

**Comment by SSoulaimane** [comment]
@JinShil isn't removing `checkAccess()` out of the scope of this PR?

**Comment by JinShil** [comment]
> isn't removing checkAccess() out of the scope of this PR?

No, I don't believe so.  When [DIP22](https://github.com/dlang/DIPs/blob/master/DIPs/archive/DIP22.md) was implemented, it had to keep both visibility checks and access checks working until the deprecation of access checks (this PR) is finished.  With visibility checks implemented, access checks are no longer needed.

The problem, as the bug above illustrates, is that the visibility checks were not fully implemented; the original authors missed that one case.  After that case is resolved, `search` will no longer return symbols that are not visible, and the code will never reach the call to `checkAccess`.

We could pull this PR as is, since the call to `accessCheck` is preventing the incorrect code from compiling, but it is not a proper solution.  We need to fix `search` so that it doesn't return symbols that aren't visible.  Maybe I can move some of the `accessCheck` code to `search`, but I'm not sure if that's the right fix.  I'll have to think about it.

**Comment by SSoulaimane** [comment]
I'm not sure you can remove that specific `checkAccess()` call. when calling `search()` on the package `imports` it should find the module `b313` since it is a child module and `isSymbolVisible()` doesn't decide whether you can access a module.

**Comment by rainers** [comment]
I hit a related problem just yesterday: https://issues.dlang.org/show_bug.cgi?id=19816

**Comment by JinShil** [comment]
I've done my best to workaround the aforementioned issue with [this commit](https://github.com/dlang/dmd/pull/9636/commits/b098efddf72e80352ccf8d45ad55803a9fef275e).  I don't think it resolves the issue that @rainers mentioned, but I don't think that should hold up this PR.

After https://github.com/vibe-d/vibe.d/pull/2304 is merged, I think this can move forward.

**Comment by wilzbach** [comment]
> After vibe-d/vibe.d#2304 is merged, I think this can move forward.

Just merged the PR. I will freeze vibe.d to master tonight, s.t. this can continue (it's just about Vibe.d's tests).

File: test/fail_compilation/fail313.d
Comment (JinShil) [reply test/fail_compilation/fail313.d]: Done.

File: changelog/remove_visibility.dd
Comment (Geod24) [inline changelog/remove_visibility.dd]: Suggest: `The deprecation phase for access checks is finished`

File: changelog/remove_visibility.dd
Comment (Geod24) [inline changelog/remove_visibility.dd]: Use backquote around the switches names

**Review by Geod24 — APPROVED** [review]
LGTM. There's still one mention of visibility, but a pre-existing, slightly wrong diagnostic should not hold this PR back.

**Comment by SSoulaimane** [comment]
@JinShil proposed cleanup of `access.d` https://github.com/SSoulaimane/dmd/commit/c72d14a4fa5c5f72cb4b4f99005b3953bd4b601d

**Comment by JinShil** [comment]
> proposed cleanup of access.d 

Done.  Thanks for that.

File: src/dmd/access.d
Comment (SSoulaimane) [inline src/dmd/access.d]: there seems to be some trailing whitespace here tha's causing the autotester failure.

**Comment by SSoulaimane** [comment]
buildkite fails are public aliases to private members.

**Comment by Geod24** [comment]
> buildkite fails are public aliases to private members.

Yeah... That should work.

**Comment by SSoulaimane** [comment]
> Yeah... That should work.
turns out alias is already resolved before calling `checkAccess` which make me feel even weirder about the old behavior of `checkAccess`.

**Comment by JinShil** [comment]
> buildkite fails are public aliases to private members.

I think that was caused by the `checkAccess` cleanup.  Those were passing before.

**Comment by SSoulaimane** [comment]
> I think that was caused by the `checkAccess` cleanup. Those were passing before.

that's how it looks like. but I cant' seem to get a working test case for it. I tried the following but it fails even with stable dmd.

```d
module a;
struct S
{
    private int m;
    public alias n = m;
 
   private int _get() { return m; }
    public alias get = _get;
}
```

```d
module b;
import a;
void main()
{
    auto s = S();
    ++s.n;
    s.get();
}
```
```shell
$ dmd b.d
b.d(6): Error: struct `a.S` member m is not accessible
b.d(7): Error: struct `a.S` member _get is not accessible
```

**Comment by JinShil** [comment]
> I tried the following but it fails even with stable dmd.

Ok, I'll try to fix that in a separate PR, and then hopfeully after that is merged and this is rebased, it will pass.

**Comment by JinShil** [comment]
I reverted the refactor of `access.d` because it introduced regressions for some projects on buildkite.  I do think it should be done, but it needs to accompany a few other changes, so I would like to do it as a followup PR.

File: src/dmd/access.d
Comment (SSoulaimane) [inline src/dmd/access.d]: This is what you need for the buildkite fails.
```diff
+    if (smember.toParent().isTemplateInstance())
+    {
+        return false; // for backwards compatibility
+    }
```

Comment (JinShil) [reply src/dmd/access.d]: Thank you.  I'd still like to do the refactoring as a followup after this PR is merged.  This PR is already quite large, and I'd like to keep the friction low.

**Comment by JinShil** [comment]
Yeah!!! All Green!!!  What a relief!!!

This PR is as good as I know how to make it at this time.  As @Geod24 mentioned there are 2 error messages that are imperfect.  They are not incorrect, but they could be better, which is like many error messages in D today.  I did my best to improve them, but it appears they will require some refactoring of DMD to get the information needed.  I don't know how to do that yet, and I think it would just add more difficulty to the review of this PR.

I will follow up this PR with @SSoulaimane's suggestion to refactor access.d.  This PR is already rather large and I would prefer not to cloud things with a refactoring.

From my point of view, this is good to go.

If you do choose to merge this, please be sure to close #7241.

Thanks @MartinNowak for the initial work, and thanks everyone else for your help along the way.

**Review by wilzbach — APPROVED** [review]
Looks very good! How about not removing the now failing tests, but just updating their error message with `AUTO_UPDATE`?

**Comment by JinShil** [comment]
I don't believe I removed any tests, but the diff doesn't show what really happened.  Some tests were just consolidated into one (e.g. the `checkImports` tests).  Others were moved (e.g. test15925.d), but the diff shows them as a deletion and an addition, rather than a rename.  Please identify a specific test if I'm missing something.

**Comment by JinShil** [comment]
BTW, this is no longer a WIP

**Comment by wilzbach** [comment]
> Edit: Also the reason the checkImports tests were consolidated is because this PR removes the -transition=checkimports flags. After doing that, the different tests we all just testing the same thing. So it made sense to only have one.

All good. Thanks! Sorry for not seeing this immediately.

**Comment by wilzbach** [comment]
BTW another nice thing about this PR is that we finally can convert the existing module imports to selective ones Phobos. In other words, we can all these selective comments:

https://github.com/dlang/phobos/blob/eb38e8e6307f2b3c27b86d2d25ddeac1b400beb5/std/algorithm/iteration.d#L74

**Review by RazvanN7 — APPROVED** [review]

**Comment by RazvanN7** [comment]
Thanks for this awesome work!

**Comment by JinShil** [comment]
And thanks to all of you for you reviews and ferrying this through the process.  It feels so good to have this done. 👍

**Comment by JinShil** [comment]
@SSoulaimane If you want to submit a PR to refactor access.d, please do.  It was your work, so it would be nice if it were signed with your handle.

**Comment by SSoulaimane** [comment]
@JinShil no it's yours just do it. I'll assist you if I can.

**Comment by FeepingCreature** [comment]
This, specifically https://github.com/dlang/dmd/pull/9636/commits/8e9b2cae2e957d95a1a47de89ba8e4227ac12132 , seems to have accidentally fixed https://issues.dlang.org/show_bug.cgi?id=19536 . You can now pass a private symbol to a template alias and call it from the template without running into errors.

I'm very happy about this, but is it intentional?

**Comment by JinShil** [comment]
> is it intentional?

No, that was not intentional, but I don't see why something like that should be prevented.  There is currently an open question about whether we should support public aliases to private symbols.  The compiler's behavior is currently inconsistent (some work -- e.g. templates -- and some don't).  I don't have a way to force design decisions from the powers that be except through the DIP process, and the DIP process is quite disproportionate when all you want is a thoughtful decision.  I think @Geod24 was going to follow up on this topic, but I'm not sure if that happened or will happen.

The current implementation proactively prevents public aliases to private symbols, so a PR to allow it should only involve deleting code.  Just by that alone, it seems like we're trying to enforce an arbitrary limitation, and I think it might be best to just to allow it, but that's not really my call to make.  I could propose it by submitting a PR, but I'm not sure if I'm prepared to fight the battle that may ensue.

**Comment by CyberShadow** [comment]
This pull request introduced a regression:
https://issues.dlang.org/show_bug.cgi?id=23991

**Pull request merged**
